src/promptflow-tools/promptflow/tools/aoai_gpt4v.py

```diff
@@ -7,12 +7,134 @@
 from promptflow._internal import ToolProvider, tool
 from promptflow.connections import AzureOpenAIConnection
 from promptflow.contracts.types import PromptTemplate
+from promptflow.exceptions import ErrorTarget, UserErrorException
+from typing import List, Dict
 
 from promptflow.tools.common import render_jinja_template, handle_openai_error, parse_chat, \
     preprocess_template_string, find_referenced_image_set, convert_to_chat_list, normalize_connection_config, \
     post_process_chat_api_response
 
 
+GPT4V_VERSION = "vision-preview"
+
+
+def _get_credential():
+    from azure.identity import DefaultAzureCredential
+    from azure.ai.ml._azure_environments import _get_default_cloud_name, EndpointURLS, _get_cloud, AzureEnvironments
+    # Support sovereign cloud cases, like mooncake, fairfax.
+    cloud_name = _get_default_cloud_name()
+    if cloud_name != AzureEnvironments.ENV_DEFAULT:
+        cloud = _get_cloud(cloud=cloud_name)
+        authority = cloud.get(EndpointURLS.ACTIVE_DIRECTORY_ENDPOINT)
+        credential = DefaultAzureCredential(authority=authority, exclude_shared_token_cache_credential=True)
+    else:
+        credential = DefaultAzureCredential()
+
+    return credential
+
+
+def _parse_resource_id(resource_id):
+    # Resource id is connection's id in following format:
+    # "/subscriptions/{sub}/resourceGroups/{rg}/providers/Microsoft.CognitiveServices/accounts/{account}"
+    split_parts = resource_id.split("/")
+    if len(split_parts) != 9:
+        raise Exception(
+            f"Connection resourceId format invalid, cur resourceId is {resource_id}."
+        )
+    sub, rg, account = split_parts[2], split_parts[4], split_parts[-1]
+
+    return sub, rg, account
+
+
+class Deployment:
+    def __init__(
+        self,
+        name: str,
+        model_name: str,
+        version: str
+    ):
+        self.name = name
+        self.model_name = model_name
+        self.version = version
+
+
+class ListDeploymentsError(UserErrorException):
+    def __init__(self, msg, **kwargs):
+        super().__init__(msg, target=ErrorTarget.TOOL, **kwargs)
+
+
+def _build_deployment_dict(item) -> Deployment:
+    model = item.properties.model
+    return Deployment(item.name, model.name, model.version)
+
+
+def list_deployment_names(
```

Comment:
is it possible to add a unit test for this function ?

Reply:
Updated.